# üìã NeonPro Coding Standards & Best Practices

**Padr√µes de c√≥digo para sistema de gest√£o em sa√∫de com qualidade ‚â•9.5/10**

## üéØ **Princ√≠pios Fundamentais**

### **KISS Principle** - Keep It Simple, Stupid

```yaml
definition: "Simplicidade √© fundamental"
core_rules:
  - Escolher solu√ß√£o mais simples que atende aos requisitos
  - Preferir c√≥digo leg√≠vel sobre otimiza√ß√µes inteligentes
  - Reduzir carga cognitiva
  - "Isso resolve o problema central sem complexidade desnecess√°ria?"
  - Usar nomenclatura clara e evitar Over-Engineering
```

### **YAGNI Principle** - You Aren't Gonna Need It

```yaml
definition: "N√£o implementar at√© ser necess√°rio"
core_rules:
  - Construir apenas o que os requisitos atuais especificam
  - Resistir a features 's√≥ por precau√ß√£o'
  - Refatorar quando requisitos emergirem
  - Focar nas user stories atuais
  - Remover c√≥digo n√£o usado imediatamente
```

### **Chain of Thought**

```yaml
definition: "Racioc√≠nio passo-a-passo expl√≠cito para precis√£o"
core_rules:
  - Quebrar problemas em passos sequenciais
  - Verbalizar processo de racioc√≠nio
  - Mostrar decis√µes intermedi√°rias
  - Questionar suposi√ß√µes
  - Validar contra requisitos
  - Cada passo segue logicamente do anterior
  - Solu√ß√£o final rastre√°vel aos requisitos
```

## üè• **Padr√µes Espec√≠ficos para Healthcare**

### **Nomenclatura Healthcare**

```typescript
// ‚úÖ CORRETO - Terminologia m√©dica clara
interface PatientRecord {
  patientId: string
  medicalRecordNumber: string
  healthInsuranceNumber: string
  emergencyContact: EmergencyContact
  lgpdConsent: ConsentStatus
}

// ‚ùå INCORRETO - Abrevia√ß√µes confusas
interface PatRec {
  id: string
  mrn: string
  ins: string
  emg: Contact
}
```

### **Error Handling Healthcare**

```typescript
// ‚úÖ CORRETO - Contexto healthcare em erros
class HealthcareError extends Error {
  constructor(
    message: string,
    public readonly healthcareContext: {
      patientId?: string
      appointmentId?: string
      clinicId?: string
      action?: string
      severity: 'low' | 'medium' | 'high' | 'critical'
    },
  ) {
    super(message,)
    this.name = 'HealthcareError'
  }
}

// ‚ùå INCORRETO - Erro gen√©rico
throw new Error('Something went wrong',)
```

## üîß **TypeScript Standards**

### **Type Definitions**

```typescript
// ‚úÖ CORRETO - Tipos espec√≠ficos e validados
interface Patient {
  readonly id: PatientId
  personalInfo: {
    fullName: string
    cpf: CPF
    dateOfBirth: Date
    gender: 'M' | 'F' | 'O'
  }
  contactInfo: {
    email: Email
    phone: PhoneNumber
    address: Address
  }
  medicalInfo: {
    bloodType?: BloodType
    allergies: Allergy[]
    medications: Medication[]
  }
  privacy: {
    lgpdConsent: boolean
    consentDate: Date
    dataRetentionUntil: Date
  }
}

// ‚ùå INCORRETO - Tipos gen√©ricos demais
interface Patient {
  id: string
  name: string
  data: any
}
```

### **Function Signatures**

```typescript
// ‚úÖ CORRETO - Fun√ß√£o pura, tipada, documentada
/**
 * Calcula pr√≥xima data dispon√≠vel considerando feriados brasileiros
 * @param baseDate - Data base para c√°lculo
 * @param excludeWeekends - Excluir fins de semana
 * @param clinicSchedule - Hor√°rios de funcionamento da cl√≠nica
 * @returns Promise<Date> - Pr√≥xima data dispon√≠vel
 */
async function calculateNextAvailableDate(
  baseDate: Date,
  excludeWeekends: boolean = true,
  clinicSchedule: ClinicSchedule,
): Promise<Date> {
  // Implementa√ß√£o...
}

// ‚ùå INCORRETO - Sem tipos, sem documenta√ß√£o
function getDate(date, options,) {
  // Implementa√ß√£o...
}
```

## ‚öõÔ∏è **React & Components Standards**

### **Component Structure**

```typescript
// ‚úÖ CORRETO - Componente bem estruturado
interface PatientCardProps {
  patient: Patient
  onEdit: (patientId: PatientId,) => void
  onDelete: (patientId: PatientId,) => Promise<void>
  readOnly?: boolean
  className?: string
  'data-testid'?: string
}

export function PatientCard({
  patient,
  onEdit,
  onDelete,
  readOnly = false,
  className,
  'data-testid': testId,
}: PatientCardProps,) {
  // Hooks no topo
  const [isDeleting, setIsDeleting,] = useState(false,)

  // Event handlers
  const handleDelete = async () => {
    setIsDeleting(true,)
    try {
      await onDelete(patient.id,)
    } catch (error) {
      // Error handling
    } finally {
      setIsDeleting(false,)
    }
  }

  // Early returns
  if (!patient) {
    return <PatientCardSkeleton />
  }

  // Main render
  return (
    <Card className={cn('patient-card', className,)} data-testid={testId}>
      {/* Component content */}
    </Card>
  )
}
```

### **Custom Hooks Pattern**

```typescript
// ‚úÖ CORRETO - Hook bem estruturado
interface UsePatientOptions {
  patientId: PatientId
  includeHistory?: boolean
  autoRefresh?: boolean
}

interface UsePatientReturn {
  patient: Patient | null
  isLoading: boolean
  error: HealthcareError | null
  refetch: () => Promise<void>
  updatePatient: (updates: Partial<Patient>,) => Promise<void>
}

export function usePatient({
  patientId,
  includeHistory = false,
  autoRefresh = false,
}: UsePatientOptions,): UsePatientReturn {
  // Estado local
  const [patient, setPatient,] = useState<Patient | null>(null,)
  const [isLoading, setIsLoading,] = useState(true,)
  const [error, setError,] = useState<HealthcareError | null>(null,)

  // L√≥gica do hook...

  return {
    patient,
    isLoading,
    error,
    refetch,
    updatePatient,
  }
}
```

## üóÑÔ∏è **Database & API Standards**

### **Database Naming**

```sql
-- ‚úÖ CORRETO - Nomenclatura consistente
CREATE TABLE patients (
  id UUID PRIMARY KEY DEFAULT gen_random_uuid(),
  full_name VARCHAR(255) NOT NULL,
  cpf VARCHAR(14) UNIQUE NOT NULL,
  date_of_birth DATE NOT NULL,
  created_at TIMESTAMPTZ DEFAULT NOW(),
  updated_at TIMESTAMPTZ DEFAULT NOW(),
  -- LGPD compliance
  data_retention_until TIMESTAMPTZ,
  lgpd_consent_date TIMESTAMPTZ
);

-- ‚ùå INCORRETO - Inconsistente
CREATE TABLE pat (
  ID int,
  nm varchar(50),
  createdAt datetime
);
```

### **API Response Pattern**

```typescript
// ‚úÖ CORRETO - Resposta padronizada
interface ApiResponse<T,> {
  success: boolean
  data?: T
  error?: {
    code: string
    message: string
    healthcareContext?: HealthcareErrorContext
  }
  meta?: {
    pagination?: PaginationInfo
    requestId: string
    timestamp: string
  }
}

// Uso
async function getPatients(): Promise<ApiResponse<Patient[]>> {
  try {
    const response = await fetch('/api/patients',)
    return await response.json()
  } catch (error) {
    return {
      success: false,
      error: {
        code: 'NETWORK_ERROR',
        message: 'Falha na conex√£o com o servidor',
      },
    }
  }
}
```

## üß™ **Testing Standards**

### **Test Structure**

```typescript
// ‚úÖ CORRETO - Teste bem estruturado
describe('PatientService', () => {
  describe('createPatient', () => {
    it('should create patient with valid data', async () => {
      // Arrange
      const validPatientData: CreatePatientData = {
        fullName: 'Jo√£o Silva',
        cpf: '123.456.789-00',
        dateOfBirth: new Date('1990-01-01',),
        email: 'joao@example.com',
      }

      // Act
      const result = await patientService.createPatient(validPatientData,)

      // Assert
      expect(result.success,).toBe(true,)
      expect(result.data,).toHaveProperty('id',)
      expect(result.data?.fullName,).toBe(validPatientData.fullName,)
    })

    it('should reject patient with invalid CPF', async () => {
      // Arrange
      const invalidPatientData = {
        ...validPatientData,
        cpf: '111.111.111-11', // CPF inv√°lido
      }

      // Act & Assert
      await expect(
        patientService.createPatient(invalidPatientData,),
      ).rejects.toThrow('CPF inv√°lido',)
    })
  })
})
```

### **Test Categories**

```typescript
// üî• CRITICAL - Business logic, APIs, financial operations
describe('Financial Calculations', () => {
  // Testes cr√≠ticos para c√°lculos financeiros
})

// ‚ö° IMPORTANT - Complex hooks, utilities, data validation
describe('Data Validation Utils', () => {
  // Testes importantes para valida√ß√µes
})

// ‚úÖ USEFUL - UI components with logic, helpers
describe('PatientCard Component', () => {
  // Testes √∫teis para componentes UI
})
```

## üìù **Documentation Standards**

### **Code Comments**

```typescript
// ‚úÖ CORRETO - Coment√°rio explicando o "porqu√™"
// Aplicamos desconto especial para pacientes SUS conforme
// regulamenta√ß√£o ANVISA 2023, artigo 15.3
const susDiscount = basePrice * 0.15

/**
 * Calcula tempo de espera m√©dio considerando prioridades m√©dicas
 *
 * Pacientes em emerg√™ncia t√™m prioridade m√°xima (0 min espera)
 * Pacientes com comorbidades t√™m prioridade alta (at√© 15 min)
 * Consultas de rotina seguem ordem de chegada
 *
 * @param appointments - Lista de agendamentos
 * @param priority - N√≠vel de prioridade m√©dica
 * @returns Tempo estimado em minutos
 */
function calculateWaitTime(
  appointments: Appointment[],
  priority: MedicalPriority,
): number {
  // Implementa√ß√£o...
}

// ‚ùå INCORRETO - Coment√°rio √≥bvio
// Incrementa o contador
counter++
```

## üîí **Security & LGPD Standards**

### **Data Privacy**

```typescript
// ‚úÖ CORRETO - Dados sens√≠veis protegidos
interface PatientPublicView {
  id: PatientId
  firstName: string // Apenas primeiro nome
  appointmentCount: number
  lastVisit: Date
  // CPF, endere√ßo, telefone omitidos
}

function getPatientPublicView(patient: Patient,): PatientPublicView {
  return {
    id: patient.id,
    firstName: patient.personalInfo.fullName.split(' ',)[0],
    appointmentCount: patient.appointments.length,
    lastVisit: patient.lastAppointment?.date ?? new Date(),
  }
}

// ‚ùå INCORRETO - Exposi√ß√£o de dados sens√≠veis
function getPatientData(patient: Patient,) {
  return patient // Exp√µe todos os dados
}
```

## üìä **Performance Standards**

### **Bundle Optimization**

```typescript
// ‚úÖ CORRETO - Import espec√≠fico
import { formatDate, } from '@neonpro/utils/date'
import { validateCPF, } from '@neonpro/utils/validation'

// ‚ùå INCORRETO - Import completo
import * as utils from '@neonpro/utils'
```

### **Component Lazy Loading**

```typescript
// ‚úÖ CORRETO - Lazy loading para componentes pesados
const PatientReportsModal = lazy(() => 
  import('@/components/PatientReportsModal').then(module => ({
    default: module.PatientReportsModal
  }))
)

// Usage
<Suspense fallback={<ReportsModalSkeleton />}>
  <PatientReportsModal />
</Suspense>
```

## ‚úÖ **Quality Checklist**

### **Pre-Commit Checklist**

- [ ] C√≥digo segue princ√≠pios KISS, YAGNI, e Chain of Thought
- [ ] Tipos TypeScript completos e precisos
- [ ] Testes escritos para l√≥gica de neg√≥cio cr√≠tica
- [ ] Documenta√ß√£o atualizada para mudan√ßas de API
- [ ] Sem vazamento de dados sens√≠veis
- [ ] Performance otimizada (imports, lazy loading)
- [ ] Nomenclatura healthcare consistente
- [ ] Error handling com contexto healthcare
- [ ] LGPD compliance verificado

### **Code Review Checklist**

- [ ] Arquitetura alinhada com padr√µes do projeto
- [ ] Seguran√ßa e privacidade de dados
- [ ] Acessibilidade (WCAG 2.1 AA+)
- [ ] Compatibilidade cross-browser
- [ ] Performance e otimiza√ß√£o
- [ ] Testes adequados e cobertura ‚â•90%

---

**Status**: ‚úÖ **ATIVO**\
**√öltima Atualiza√ß√£o**: 2025-01-08\
**Qualidade Target**: ‚â•9.5/10
